### PR TITLE
Adding quaternion to yaw, pitch, roll control angles math

### PR DESCRIPTION
**Pull request by Timot05**
The sensor fusion should produce a quaternion representing the attitude of the controller. From there, we should be able to reset the quaternion's position to zero once we arm the controller. The yaw, pitch and roll angles should be extracted from the quaternion and sent over to the transceiver.

### CODE REVIEW

**Comment by MaEtUgR** [comment]
If the input is the quaternion representing the rotation from the body frame of the controller to the local world frame of the person holding it then it's start by decomposing the quaternion into a tilt vector or quaternion and yaw like so: https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mc_att_control/AttitudeControl/AttitudeControl.cpp#L59-L62
The resetting is to account for the yaw/heading that the controller is initially held in right? The tilt then easily to how much roll and pitch input is given.

Are the calculations added in this pr what you previously used for flying?

File: firmware/debug/teapot.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    roll  = math.atan2(2.0 * (q[0] * q[1] + q[2] * q[3]), q[0] * q[0] - q[1] * q[1] - q[2] * q[2] + q[3] * q[3])
+    pitch *= 180.0 / math.pi
+    yaw   *= 180.0 / math.pi
+    yaw   -= -0.13  # Declination at Chandrapur, Maharashtra is - 0 degress 13 min
+    roll  *= 180.0 / math.pi
+    return [yaw, pitch, roll]
+
+
+if __name__ == '__main__':
+    main()
```
Comment (MaEtUgR) [inline firmware/debug/teapot.py:215]: no newline

File: firmware/src/main.cpp [hunk 2]
```diff
@@ -98,6 +105,76 @@ void quaternionToEulerGI(sh2_GyroIntegratedRV_t *rotational_vector, euler_t *ypr
...
+	// ypr->yaw =
+
+    //yaw rotation around x
+    //pitch rotation around z
+    //roll rotation around y
+	// The quaternion order and sign has to be changed to ensure that the YPR values are correct
+    float q0 = quaternion->w; //w
+    float q1 = quaternion->z; //x
+    float q2 = - quaternion->x; //y
+    float q3 = quaternion->y; //z
```
Comment (MaEtUgR) [inline firmware/src/main.cpp:129]: What quaternion convention is the input in?
https://www.iri.upc.edu/people/jsola/JoanSola/objectes/notes/kinematics.pdf Chapter 3
e.g. Eigen PX4 and so on use Hamilton convention.

File: firmware/src/main.cpp [hunk 3]
```diff
@@ -98,6 +105,76 @@ void quaternionToEulerGI(sh2_GyroIntegratedRV_t *rotational_vector, euler_t *ypr
...
+    float fAngle;
+    float fInvSqrt_q0q0_q1q1 = 1.0f/sqrtf(q0 * q0 + q1 * q1);
+
+    float r11, r12;
+
+    float xY = 0;
+    float xP = 0;
+    float xR = 0;
+
+    /* Yaw extraction */
```
Comment (MaEtUgR) [inline firmware/src/main.cpp:140]: Maybe it comes out to be the same as the decomposition I was commenting about. I have to run through once to check.

File: firmware/src/main.cpp (hunk 2 above)
Comment (Timot05) [reply firmware/src/main.cpp:129]: Happy to adopt any standard here! Open to your suggestions. We should check what standard comes out of the BNO086 from Bosch and convert it to whatever standard we will use for the remote.

I've been using [this script](https://github.com/thecountoftuscany/PyTeapot-Quaternion-Euler-cube-rotation) which uses another standard too, that's why it's a bit of a mess currently.

**Comment by Timot05** [comment]
@MaEtUgR Yes, the calculations are the ones that were used previously. When the controller is armed, the quaternion used to calculate the YPR angles needs to be reset to zero (haven't added the math for that yet). From there, we can input it into the `quaternionToControlAngles()` function and get the control angles out.

**Pull request closed without merging**